lib/xnvme_be_macos_driverkit_async.c

```diff
@@ -0,0 +1,219 @@
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)queue->base.dev->be.state;
+	uint64_t qid[2] = {q->id, q->id}; // Submission queue, completion queue
+	int reaped = 0;
+	int queue_ret;
+	NvmeSubmitCmd nvme_cmd;
+	kern_return_t ret;
+	struct xnvme_cmd_ctx *ctx;
+
+	ret = IOConnectCallScalarMethod(state->device_connection, NVME_POKE, qid, 2, NULL, 0);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallScalarMethod(); ret(0x%x), msg(%s)", ret,
+			    mach_error_string(ret));
+		return -EIO;
+	}
+	while ((queue_ret = queue_dequeue(q->cq_queue, &nvme_cmd)) == 0) {
+		ctx = (struct xnvme_cmd_ctx *)nvme_cmd.backend_opaque;
+		memcpy(&ctx->cpl, nvme_cmd.cpl, sizeof(struct xnvme_spec_cpl));
+		ctx->async.cb(ctx, ctx->async.cb_arg);
+
+		reaped += 1;
+		queue->base.outstanding -= 1;
+	}
+
+	if (queue_ret == QUEUE_ERROR) {
+		XNVME_DEBUG("FAILED: queue_ret == QUEUE_ERROR");
+		return -EIO;
+	}
+
+	return reaped;
+}
+
+int
+xnvme_be_macos_driverkit_async_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nbytes,
+				      void *mbuf, size_t mbuf_nbytes)
+{
+	struct xnvme_queue_macos_driverkit *q =
+		(struct xnvme_queue_macos_driverkit *)ctx->async.queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)ctx->dev->be.state;
+	NvmeSubmitCmd cmd = {0};
+	int queue_ret;
+	struct buffer *buf;
+	uint64_t _buf_base_offset;
+
+	cmd.queue_id = state->qid_sync;
+
+	switch (ctx->cmd.common.opcode) {
+	case XNVME_SPEC_NVM_OPC_READ:
+		break;
+
+	case XNVME_SPEC_FS_OPC_READ:
+		ctx->cmd.nvm.slba = ctx->cmd.nvm.slba >> ctx->dev->geo.ssw;
+		ctx->cmd.common.opcode = XNVME_SPEC_NVM_OPC_READ;
+		break;
+
+	case XNVME_SPEC_NVM_OPC_WRITE:
+		break;
+
+	case XNVME_SPEC_FS_OPC_WRITE:
+		ctx->cmd.nvm.slba = ctx->cmd.nvm.slba >> ctx->dev->geo.ssw;
+		ctx->cmd.common.opcode = XNVME_SPEC_NVM_OPC_WRITE;
+		break;
+	}
+
+	if (dbuf) {
+		buf = buffer_find(state->buffers, (uint64_t)dbuf);
+		if (!buf) {
+			return -EINVAL;
+		}
+		_buf_base_offset = ((uint64_t)dbuf) - buf->vaddr;
+		cmd.dbuf_nbytes = dbuf_nbytes;
+		cmd.dbuf_offset = _buf_base_offset;
+		cmd.dbuf_token = buf->vaddr;
+		XNVME_DEBUG("INFO: buf token: %llx: %llx, %llx", (uint64_t)buf->vaddr,
+			    (uint64_t)dbuf, cmd.dbuf_offset);
+	}
+	if (mbuf) {
+		buf = buffer_find(state->buffers, (uint64_t)mbuf);
+		if (!buf) {
+			XNVME_DEBUG("FAILED: buffer_find()");
+			return -EINVAL;
+		}
+		_buf_base_offset = ((uint64_t)mbuf) - buf->vaddr;
+		cmd.mbuf_nbytes = mbuf_nbytes;
+		cmd.mbuf_offset = _buf_base_offset;
+		cmd.mbuf_token = buf->vaddr;
+		XNVME_DEBUG("INFO: buf token: %llx: %llx, %llx", (uint64_t)buf->vaddr,
+			    (uint64_t)mbuf, cmd.mbuf_offset);
+	}
+	memcpy(cmd.cmd, &ctx->cmd, sizeof(struct xnvme_spec_cmd));
+	cmd.backend_opaque = (uint64_t)ctx;
+
+	queue_ret = queue_enqueue(q->sq_queue, &cmd);
+	if (queue_ret == QUEUE_FULL) {
+		return -EBUSY;
+	} else if (queue_ret) {
```

Comment:
Should have ``XNVME_DEBUG("FAILED: ...")`` on the error-path here.

Reply:
Done.